sdk/core/Azure.Core/src/Shared/OperationInternal.cs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+
+namespace Azure.Core
+{
+    internal class OperationInternal<TResult>
+    {
+        private const string RetryAfterHeaderName = "Retry-After";
+
+        private readonly object SetStateLock = new object();
+
+        private readonly IOperation<TResult> _operation;
+
+        private TResult _value;
+
+        private RequestFailedException _operationFailedException;
+
+        public OperationInternal(ClientDiagnostics clientDiagnostics, IOperation<TResult> operation)
+        {
+            _operation = operation;
+            OperationTypeName = operation.GetType().Name;
+            ClientDiagnostics = clientDiagnostics;
+            DefaultPollingInterval = TimeSpan.FromSeconds(1);
+            ScopeAttributes = new Dictionary<string, string>();
+        }
+
+        public IDictionary<string, string> ScopeAttributes { get; }
+
+        public bool HasValue { get; private set; }
+
+        public bool HasCompleted { get; set; }
+
+        public TResult Value
+        {
+            get
+            {
+                if (HasValue)
+                {
+                    return _value;
+                }
+                else if (OperationFailedException != null)
+                {
+                    throw OperationFailedException;
+                }
+                else
+                {
+                    throw new InvalidOperationException("The operation has not completed yet.");
+                }
+            }
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(Value));
+                }
+
+                _value = value;
+                HasValue = true;
+            }
+        }
+
+        public Response RawResponse { get; set; }
+
+        public string OperationTypeName { get; set; }
+
+        public TimeSpan DefaultPollingInterval { get; set; }
+
+        private ClientDiagnostics ClientDiagnostics { get; }
+
+        private RequestFailedException OperationFailedException
+        {
+            get => _operationFailedException;
+            set
+            {
+                if (value is null)
+                {
+                    throw new ArgumentNullException(nameof(value));
+                }
+
+                _operationFailedException = value;
```

Comment:
Do we need a property?

Reply:
I was using it for the null check. Since we have a default exception now, not anymore.